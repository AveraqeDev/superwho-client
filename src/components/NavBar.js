import React, { useState, useEffect, useContext } from 'react';
import { Link, withRouter } from 'react-router-dom';
import { CSSTransition } from 'react-transition-group';

import TokenService from '../services/token-service';
import { UserContext } from '../contexts/UserContext';

import '../styles/NavBar.css';

const Branding = React.memo(function Branding() {
  return (
    <Link className="Branding" to="/">SuperWho?</Link>
  );
});

const Burger = React.memo(function Burger({ open, setOpen }) {
  return (
    <button
    aria-label='Burger Menu'
    className='Burger'
    onClick={() => setOpen(!open)}
    >
      <div className={open ? 'burger_open' : ''} />
      <div className={open ? 'burger_open' : ''} />
      <div className={open ? 'burger_open' : ''} />
    </button>
  )
});

const Menu = React.memo(function Menu({ open, setOpen, onLogout, location }) {
  const { pathname } = location;

  const { user } = useContext(UserContext);

  return (
    <div className="Menu">
      <Link className={`Menu__Link${pathname === '/' ? ' Menu__Link-selected' : ''}`} to='/' onClick={() => setOpen(!open)}>Search</Link>
      {(user.id)
        ? <Link className={`Menu__Link${pathname === '/favorites' ? ' Menu__Link-selected' : ''}`} to='/favorites' onClick={() => setOpen(!open)}>Favorites</Link>
        : ''
      }
      {(user.id)
        ? <Link className={`Menu__Link${pathname === '/profile' ? ' Menu__Link-selected' : ''}`} to='/profile' onClick={() => setOpen(!open)}>Profile</Link>
        : <Link className={`Menu__Link${pathname === '/login' ? ' Menu__Link-selected' : ''}`} to='/login' onClick={() => setOpen(!open)}>Login</Link>
      }
      {(user.id)
        ? <button className="Menu__Link" onClick={onLogout}>Logout</button>
        : <Link className={`Menu__Link${pathname === '/register' ? ' Menu__Link-selected' : ''}`} to='/register' onClick={() => setOpen(!open)}>Register</Link>
      }
    </div>
  );
});

const UserAccount = React.memo(function UserAccount({ open, setOpen }) {
  const { user } = useContext(UserContext);

  if(user) {
    return (
      <div className="UserAccount">
        <p onClick={() => setOpen(!open)}>{user.username} <span className="DownArrow">â–¼</span></p>
      </div>
    );
  } else {
    return (
      <div className="UserAccount">
        <Link className="UserAccount__Link" to="/login">Login</Link>
        <Link className="UserAccount__Link" to="/register">Register</Link>
      </div>
    );
  }
});

const UserAccountMenu = React.memo(function UserAccountMenu({ open, setOpen, onLogout }) {
  return (
    <div className="UserAccountMenu">
      <Link className="UserAccountMenu__Link" to="/profile" onClick={() => setOpen(!open)}>Profile</Link>
      <hr className="seperator"></hr>
      <button className="UserAccountMenu__Link" onClick={onLogout}>Logout</button>
    </div>
  );
});

const Sidebar = React.memo(function Sidebar(props) {
  const { pathname } = props.location;

  const { user } = useContext(UserContext);

  return (
    <div className="Menu">
      <Link className={`Menu__Link${pathname === '/' ? ' Menu__Link-selected' : ''}`} to='/'>Search</Link>
      {(user)
        ? <Link className={`Menu__Link${pathname === '/favorites' ? ' Menu__Link-selected' : ''}`} to='/favorites'>Favorites</Link>
        : ''
      }
    </div>
  );
});

const NavBar = (props) => {
  const [navOpen, setNavOpen] = useState(false);
  const [accountOpen, setAccountOpen] = useState(false);
  const [windowWidth, setWindowWidth] = useState(0);
  const resizeWindow = () => {
    setWindowWidth(window.innerWidth);
  };

  const { setUser } = useContext(UserContext);

  useEffect(() => {
    resizeWindow();
    window.addEventListener("resize", resizeWindow);
    return () => window.removeEventListener("resize", resizeWindow);
  }, []);

  const onLogout = () => {
    setNavOpen(false);
    setAccountOpen(false);
    TokenService.clearAuthToken();
    setUser(null);
    props.history.push('/');
  };

  if(props.location.pathname === '/login' || props.location.pathname === '/register') {
    return '';
  }

  if(windowWidth <= 760) {
    return (
      <nav className="NavBar">
        <Burger open={navOpen} setOpen={setNavOpen} />
        <CSSTransition
          in={navOpen}
          timeout={300}
          classNames='Menu'
          unmountOnExit
        >
          <Menu open={navOpen} setOpen={setNavOpen} onLogout={onLogout} location={props.location} />
        </CSSTransition>
        <Branding />
      </nav>
    );
  } else {
    return (
      <nav className="NavBar">
        <Sidebar location={props.location} />
        <Branding />
        <UserAccount open={accountOpen} setOpen={setAccountOpen} />
        <CSSTransition
          in={accountOpen}
          timeout={300}
          classNames='AccountMenu'
          unmountOnExit
        >
          <UserAccountMenu open={accountOpen} setOpen={setAccountOpen} onLogout={onLogout} />
        </CSSTransition>
      </nav>
    )
  }
};

export default withRouter(NavBar);