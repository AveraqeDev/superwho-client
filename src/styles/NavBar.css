/* Main NavBar */

.NavBar {
  background: #2A9BE4;
  display: flex;
  justify-content: space-between;
  align-items: center;
  filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25));
  height: 65px;
}

.Branding {
  font-style: normal;
  font-weight: bold;
  font-size: 40px;
  text-align: center;
  color: #FFFFFF;
  margin-left: 50px;
  text-decoration: none;
}

/* UserAccount Dropdown */

.UserAccount {
  font-style: normal;
  font-weight: bold;
  font-size: 22px;
  text-align: center;
  color: #FFFFFF;
  margin-right: 50px;
  cursor: pointer;
}

.UserAccount:hover {
  color: #CFCFCF;
}

.DownArrow {
  font-size: medium;
}

.UserAccount__Link {
  text-decoration: none;
  color: #FFFFFF;
  padding-left: 1.5rem;
  transition: all 0.2s linear;
}

.UserAccount__Link:hover {
  color: #CFCFCF;
}

.UserAccountMenu {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  position: absolute;
  width: 200px;
  right: 35px;
  top: 55px;
  background: #FFFFFF;
  border-radius: 5px;
  padding: .7rem 0;
  text-align: center;
}

.UserAccountMenu__Link {
  border: none;
  background: #FFFFFF;
  border-radius: 5px;
  font-weight: bold;
  font-size: 18px;
  line-height: 24px;
  color: #A7A8BA;
  width: 100%;
  height: 100%;
  cursor: pointer;
  padding: 5px;
  text-decoration: none;
  transition: all 0.2s linear;
}

.UserAccountMenu__Link:hover {
  color: #CFCFCF;
}

.UserAccountMenu__Link:focus {
  outline: none;
}

.AccountMenu-enter {
  opacity: 0;
}

.AccountMenu-enter-active {
  opacity: 1;
  transition: opacity 0.3s ease-in-out;
}

.AccountMenu-exit-active {
  opacity: 0;
  transition: opacity 0.3s ease-in-out;
}

.seperator {
  width: 100%;
  background: #A7A8BA;
}

/* Burger */

.Burger {
  position: absolute;
  left: 2rem;
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  width: 2rem;
  height: 2rem;
  background: transparent;
  border: none;
  cursor: pointer;
  padding: 0;
  z-index: 10;
}

.Burger:focus {
  outline: none;
}

.Burger:hover > div {
  background: #CFCFCF;
}

.Burger > div {
  width: 2rem;
  height: 0.25rem;
  background: #FFFFFF;
  border-radius: 10px;
  transition: all 0.2s linear;
  position: relative;
  transform-origin: 1px;
}

.burger_open:first-child {
  transform: rotate(45deg);
}

.burger_open:nth-child(2) {
  opacity: 0;
  transform: translateX(-20px);
}

.burger_open:nth-child(3) {
  transform: rotate(-45deg);
}

/* Menu */

.Menu {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  background: #404853;
  text-align: center;
  position: absolute;
  top: 65px;
  left: 0;
  width: 12%;
  min-width: 130px;
  height: calc(100% - 65px);
  z-index: 1;
}

.Menu__Link {
  font-size: 24px;
  font-weight: bold;
  color: #FFFFFF;
  transition: all 0.2s linear;
  background: none;
  border: none;
  text-decoration: none;
  cursor: pointer;
  padding: 1.9rem 0;
}

.Menu__Link-selected {
  background: #233038;
}

.Menu__Link:hover {
  color: #CFCFCF;
  opacity: 0.8;
}

.Menu-enter {
  transform: translateX(-100%);
}

.Menu-enter-active {
  transform: translateX(0);
  transition: transform 0.3s ease-in-out;
}

.Menu-exit-active {
  transform: translateX(-100%);
  transition: transform 0.3s ease-in-out;
}

/* Media Queries */

@media screen and (max-width: 760px) {
  .NavBar {
    justify-content: center;
  }

  .Menu {
    width: 20%;
  }
}

@media screen and (max-width: 420px) {
  .Menu {
    width: 200px;
  }
}