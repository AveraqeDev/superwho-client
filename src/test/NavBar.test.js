import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import NavBar, { Branding, Burger, Menu, UserAccount, UserAccountMenu, Sidebar } from '../components/NavBar';


describe('NavBar', () => {
  describe('<Branding />', () => {
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <MemoryRouter>
          <Branding />
        </MemoryRouter>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <Branding />
          </MemoryRouter>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<Burger />', () => {
    const testData = {
      open: false,
      setOpen: () => {}
    };
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<Burger open={testData.open} setOpen={testData.setOpen} />, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <Burger open={testData.open} setOpen={testData.setOpen} />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<Menu />', () => {
    const testData = {
      open: false,
      setOpen: () => {},
      onLogout: () => {},
      location: ''
    };
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <MemoryRouter>
          <Menu open={testData.open} setOpen={testData.setOpen} onLogout={testData.onLogout} location={testData.location} />
        </MemoryRouter>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <Menu open={testData.open} setOpen={testData.setOpen} onLogout={testData.onLogout} location={testData.location} />
          </MemoryRouter>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<UserAccount />', () => {
    const testData = {
      open: false,
      setOpen: () => {}
    };
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <MemoryRouter>
          <UserAccount open={testData.open} setOpen={testData.setOpen} />
        </MemoryRouter>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <UserAccount open={testData.open} setOpen={testData.setOpen} />
          </MemoryRouter>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<UserAccountMenu />', () => {
    const testData = {
      open: false,
      setOpen: () => {},
      onLogout: () => {}
    };
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <MemoryRouter>
          <UserAccountMenu open={testData.open} setOpen={testData.setOpen} onLogout={testData.onLogout} />
        </MemoryRouter>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <UserAccountMenu open={testData.open} setOpen={testData.setOpen} onLogout={testData.onLogout} />
          </MemoryRouter>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<Sidebar />', () => {
    const testData = {
      location: {
        pathname: ''
      }
    };
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <MemoryRouter>
          <Sidebar location={testData.location} />
        </MemoryRouter>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <Sidebar location={testData.location} />
          </MemoryRouter>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });

  describe('<NavBar />', () => {
    // Smoke test
    it('renders without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(
        <MemoryRouter>
          <NavBar />
        </MemoryRouter>, div);
      ReactDOM.unmountComponentAtNode(div);
    });

    //Snapshot test
    it('renders the UI as expected', () => {
      const tree = renderer
        .create(
          <MemoryRouter>
            <NavBar />
          </MemoryRouter>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    })
  });
});