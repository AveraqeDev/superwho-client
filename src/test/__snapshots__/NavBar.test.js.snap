// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar <Branding /> renders the UI as expected 1`] = `
<a
  className="Branding"
  href="/"
  onClick={[Function]}
>
  SuperWho?
</a>
`;

exports[`NavBar <Burger /> renders the UI as expected 1`] = `
<button
  aria-label="Burger Menu"
  className="Burger"
  onClick={[Function]}
>
  <div
    className=""
  />
  <div
    className=""
  />
  <div
    className=""
  />
</button>
`;

exports[`NavBar <Menu /> renders the UI as expected 1`] = `
<div
  className="Menu"
>
  <a
    className="Menu__Link"
    href="/"
    onClick={[Function]}
  >
    Search
  </a>
  <a
    className="Menu__Link"
    href="/favorites"
    onClick={[Function]}
  >
    Favorites
  </a>
  <a
    className="Menu__Link"
    href="/profile"
    onClick={[Function]}
  >
    Profile
  </a>
  <button
    className="Menu__Link"
    onClick={[Function]}
  >
    Logout
  </button>
</div>
`;

exports[`NavBar <NavBar /> renders the UI as expected 1`] = `
<nav
  className="NavBar"
>
  <button
    aria-label="Burger Menu"
    className="Burger"
    onClick={[Function]}
  >
    <div
      className=""
    />
    <div
      className=""
    />
    <div
      className=""
    />
  </button>
  <a
    className="Branding"
    href="/"
    onClick={[Function]}
  >
    SuperWho?
  </a>
</nav>
`;

exports[`NavBar <Sidebar /> renders the UI as expected 1`] = `
<div
  className="Menu"
>
  <a
    className="Menu__Link"
    href="/"
    onClick={[Function]}
  >
    Search
  </a>
  <a
    className="Menu__Link"
    href="/favorites"
    onClick={[Function]}
  >
    Favorites
  </a>
</div>
`;

exports[`NavBar <UserAccount /> renders the UI as expected 1`] = `
<div
  className="UserAccount"
>
  <p
    onClick={[Function]}
  >
     
    <span
      className="DownArrow"
    >
      â–¼
    </span>
  </p>
</div>
`;

exports[`NavBar <UserAccountMenu /> renders the UI as expected 1`] = `
<div
  className="UserAccountMenu"
>
  <a
    className="UserAccountMenu__Link"
    href="/profile"
    onClick={[Function]}
  >
    Profile
  </a>
  <hr
    className="seperator"
  />
  <button
    className="UserAccountMenu__Link"
    onClick={[Function]}
  >
    Logout
  </button>
</div>
`;
